File: src/ee/vungle/private/VungleBridge.swift
//
//  Vungle.swift
//  Pods
//
//  Created by eps on 6/25/20.
//

import Foundation
import RxSwift

private let kTag = "\(VungleBridge.self)"
private let kPrefix = "VungleBridge"
private let kInitialize = "\(kPrefix)Initialize"
private let kHasRewardedAd = "\(kPrefix)HasRewardedAd"
private let kLoadRewardedAd = "\(kPrefix)LoadRewardedAd"
private let kShowRewardedAd = "\(kPrefix)ShowRewardedAd"
private let kOnLoaded = "\(kPrefix)OnLoaded"
private let kOnFailedToLoad = "\(kPrefix)OnFailedToLoad"
private let kOnFailedToShow = "\(kPrefix)OnFailedToShow"
private let kOnClicked = "\(kPrefix)OnClicked"
private let kOnClosed = "\(kPrefix)OnClosed"

@objc(EEVungleBridge)
public class VungleBridge: NSObject, IPlugin, VungleSDKDelegate {
    private let _bridge: IMessageBridge
    private let _logger: ILogger
    private let _sdk = VungleSDK.shared()
    private var _initializing = false
    private var _initialized = false
    private var _initializationAwaiter: (() -> Void)?
    private var _rewarded = false
    private var _loadingAdIds: Set<String> = []
    private var _loadedAdIds: Set<String> = []

    public required init(_ bridge: IMessageBridge, _ logger: ILogger) {
        _bridge = bridge
        _logger = logger
        super.init()
        registerHandlers()
    }

    public func destroy() {
        deregisterHandlers()
        Thread.runOnMainThread {
            if !self._initialized {
                return
            }
            self._sdk.delegate = nil
        }
    }

    func registerHandlers() {
        _bridge.registerAsyncHandler(kInitialize) { message, resolver in
            let dict = EEJsonUtils.convertString(toDictionary: message)
            guard let appId = dict["appId"] as? String else {
                assert(false, "Invalid argument")
                return
            }
            self.initialize(appId)
                .subscribe(
                    onSuccess: {
                        result in resolver(Utils.toString(result))
                    }, onError: {
                        _ in resolver(Utils.toString(false))
                    })
        }
        _bridge.registerHandler(kHasRewardedAd) { message in
            Utils.toString(self.hasRewardedAd(message))
        }
        _bridge.registerHandler(kLoadRewardedAd) { message in
            self.loadRewardedAd(message)
            return ""
        }
        _bridge.registerHandler(kShowRewardedAd) { message in
            self.showRewardedAd(message)
            return ""
        }
    }

    func deregisterHandlers() {
        _bridge.deregisterHandler(kInitialize)
        _bridge.deregisterHandler(kHasRewardedAd)
        _bridge.deregisterHandler(kLoadRewardedAd)
        _bridge.deregisterHandler(kShowRewardedAd)
    }

    func initialize(_ appId: String) -> Single<Bool> {
        return Single<Bool>.create { single in
            Thread.runOnMainThread {
                if self._initializing {
                    single(.success(false))
                    return
                }
                if self._initialized {
                    single(.success(true))
                    return
                }
                self._initializing = true
                self._sdk.delegate = self
                do {
                    self._initializationAwaiter = {
                        single(.success(true))
                    }
                    try self._sdk.start(withAppId: appId)
                } catch {
                    self._logger.debug("\(kTag): \(#function): \(error.localizedDescription)")
                    return
                }
            }
            return Disposables.create()
        }
        .subscribeOn(MainScheduler())
    }

    func hasRewardedAd(_ adId: String) -> Bool {
        return _loadedAdIds.contains(adId)
    }

    func loadRewardedAd(_ adId: String) {
        Thread.runOnMainThread {
            do {
                self._loadingAdIds.insert(adId)
                try self._sdk.loadPlacement(withID: adId)
            } catch {
                self._loadingAdIds.remove(adId)
                self._bridge.callCpp(kOnFailedToLoad, EEJsonUtils.convertDictionary(toString: [
                    "ad_id": adId,
                    "message": error.localizedDescription
                ]))
                return
            }
            // OK.
        }
    }

    func showRewardedAd(_ adId: String) {
        Thread.runOnMainThread {
            guard let rootView = Utils.getCurrentRootViewController() else {
                assert(false, "Root view is null")
                return
            }
            self._rewarded = false
            do {
                try self._sdk.playAd(rootView, options: nil, placementID: adId)
            } catch {
                self._bridge.callCpp(kOnFailedToShow, EEJsonUtils.convertDictionary(toString: [
                    "ad_id": adId,
                    "message": error.localizedDescription
                ]))
                return
            }
            // OK.
        }
    }

    public func vungleSDKDidInitialize() {
        _logger.debug("\(kTag): \(#function)")
        _initializing = false
        _initialized = false
        if let awaiter = _initializationAwaiter {
            awaiter()
        }
    }

    public func vungleSDKFailedToInitializeWithError(_ error: Error) {
        _logger.debug("\(kTag): \(#function): \(error.localizedDescription)")
        _initializing = false
    }

    public func vungleAdPlayabilityUpdate(_ isAdPlayable: Bool, adId: String?, error: Error?) {
        _logger.debug("\(kTag): \(#function): playable = \(isAdPlayable) id = \(adId ?? "") reason \(error?.localizedDescription ?? "")")
        if let adId = adId {
            if isAdPlayable {
                _loadedAdIds.insert(adId)
            }
            if _loadingAdIds.contains(adId) {
                assert(isAdPlayable)
                _loadingAdIds.remove(adId)
                _bridge.callCpp(kOnLoaded, EEJsonUtils.convertDictionary(toString: [
                    "ad_id": adId
                ]))
                return
            }
        }
        if error != nil {
            assert(adId == nil)
            return
        }
        // Mediation ???
    }

    public func vungleWillShowAd(forPlacementID adId: String?) {
        _logger.debug("\(kTag): \(#function): \(adId ?? "")")
    }

    public func vungleDidShowAd(forPlacementID adId: String?) {
        _logger.debug("\(kTag): \(#function): \(adId ?? "")")
    }

    public func vungleTrackClick(forPlacementID adId: String?) {
        _logger.debug("\(kTag): \(#function): \(adId ?? "")")
        _bridge.callCpp(kOnClicked, adId ?? "")
    }

    public func vungleWillLeaveApplication(forPlacementID adId: String?) {
        _logger.debug("\(kTag): \(#function): \(adId ?? "")")
    }

    public func vungleRewardUser(forPlacementID adId: String?) {
        _logger.debug("\(kTag): \(#function): \(adId ?? "")")
        _rewarded = true
    }

    public func vungleWillCloseAd(forPlacementID adId: String) {
        _logger.debug("\(kTag): \(#function): \(adId)")
    }

    public func vungleDidCloseAd(forPlacementID adId: String) {
        _logger.debug("\(kTag): \(#function): \(adId)")
        _loadedAdIds.remove(adId)
        _bridge.callCpp(kOnClosed, EEJsonUtils.convertDictionary(toString: [
            "ad_id": adId,
            "rewarded": _rewarded
        ]))
    }
}

File: src/ee/vungle/private/VungleBridge.cpp
#include "ee/vungle/private/VungleBridge.hpp"

#include <ee/ads/internal/GuardedRewardedAd.hpp>
#include <ee/ads/internal/IAsyncHelper.hpp>
#include <ee/ads/internal/MediationManager.hpp>
#include <ee/core/IMessageBridge.hpp>
#include <ee/core/Logger.hpp>
#include <ee/core/PluginManager.hpp>
#include <ee/core/Task.hpp>
#include <ee/core/Thread.hpp>
#include <ee/core/Utils.hpp>
#include <ee/nlohmann/json.hpp>

#include "ee/vungle/private/VungleRewardedAd.hpp"

namespace ee {
namespace core {
template <>
std::shared_ptr<IVungle>
PluginManager::createPluginImpl(IMessageBridge& bridge) {
    if (not addPlugin(Plugin::Vungle)) {
        return nullptr;
    }
    return std::make_shared<vungle::Bridge>(bridge);
}
} // namespace core

namespace vungle {
namespace {
// clang-format off
const std::string kPrefix    = "VungleBridge";

const auto kInitialize     = kPrefix + "Initialize";

const auto kHasRewardedAd  = kPrefix + "HasRewardedAd";
const auto kLoadRewardedAd = kPrefix + "LoadRewardedAd";
const auto kShowRewardedAd = kPrefix + "ShowRewardedAd";

const auto kOnLoaded       = kPrefix + "OnLoaded";
const auto kOnFailedToLoad = kPrefix + "OnFailedToLoad";
const auto kOnFailedToShow = kPrefix + "OnFailedToShow";
const auto kOnClicked      = kPrefix + "OnClicked";
const auto kOnClosed       = kPrefix + "OnClosed";
// clang-format on
} // namespace

using Self = Bridge;

Self::Bridge(IMessageBridge& bridge)
    : bridge_(bridge)
    , logger_(Logger::getSystemLogger()) {
    logger_.debug(__PRETTY_FUNCTION__);
    auto&& mediation = ads::MediationManager::getInstance();
    rewardedAdDisplayer_ = mediation.getRewardedAdDisplayer();

    bridge_.registerHandler(
        [this](const std::string& message) {
            Thread::runOnLibraryThread([this, message] { //
                auto json = nlohmann::json::parse(message);
                onLoaded(json["ad_id"]);
            });
            return "";
        },
        kOnLoaded);
    bridge_.registerHandler(
        [this](const std::string& message) {
            Thread::runOnLibraryThread([this, message] { //
                auto json = nlohmann::json::parse(message);
                onFailedToLoad(json["ad_id"], json["message"]);
            });
            return "";
        },
        kOnFailedToLoad);
    bridge_.registerHandler(
        [this](const std::string& message) {
            Thread::runOnLibraryThread([this, message] { //
                auto json = nlohmann::json::parse(message);
                onFailedToShow(json["ad_id"], json["message"]);
            });
            return "";
        },
        kOnFailedToShow);
    bridge_.registerHandler(
        [this](const std::string& message) {
            Thread::runOnLibraryThread([this, message] { //
                onClicked(message);
            });
            return "";
        },
        kOnClicked);
    bridge_.registerHandler(
        [this](const std::string& message) {
            Thread::runOnLibraryThread([this, message] { //
                auto json = nlohmann::json::parse(message);
                onClosed(json["ad_id"], json["rewarded"]);
            });
            return "";
        },
        kOnClosed);
}

Self::~Bridge() = default;

void Self::destroy() {
    logger_.debug(__PRETTY_FUNCTION__);
    bridge_.deregisterHandler(kOnLoaded);
    bridge_.deregisterHandler(kOnFailedToLoad);
    bridge_.deregisterHandler(kOnFailedToShow);
    bridge_.deregisterHandler(kOnClosed);
    PluginManager::removePlugin(Plugin::Vungle);
}

Task<bool> Self::initialize(const std::string& appId) {
    logger_.debug("%s: appId = %s", __PRETTY_FUNCTION__, appId.c_str());
    nlohmann::json json;
    json["appId"] = appId;
    auto response = co_await bridge_.callAsync(kInitialize, json.dump());
    co_return core::toBool(response);
}

std::shared_ptr<IRewardedAd> Self::createRewardedAd(const std::string& adId) {
    logger_.debug("%s: adId = %s", __PRETTY_FUNCTION__, adId.c_str());
    auto iter = rewardedAds_.find(adId);
    if (iter != rewardedAds_.cend()) {
        return iter->second.strong;
    }
    auto weak = new RewardedAd(logger_, rewardedAdDisplayer_, this, adId);
    auto strong = std::make_shared<ads::GuardedRewardedAd>(
        std::shared_ptr<RewardedAd>(weak));
    rewardedAds_.try_emplace(adId, strong, weak);
    return strong;
}

bool Self::destroyRewardedAd(const std::string& adId) {
    logger_.debug("%s: adId = %s", __PRETTY_FUNCTION__, adId.c_str());
    auto iter = rewardedAds_.find(adId);
    if (iter == rewardedAds_.cend()) {
        return false;
    }
    rewardedAds_.erase(iter);
    return true;
}

bool Self::hasRewardedAd(const std::string& adId) const {
    auto result = bridge_.call(kHasRewardedAd, adId);
    return core::toBool(result);
}

void Self::loadRewardedAd(const std::string& adId) const {
    logger_.debug("%s: adId = %s", __PRETTY_FUNCTION__, adId.c_str());
    bridge_.call(kLoadRewardedAd, adId);
}

void Self::showRewardedAd(const std::string& adId) {
    logger_.debug("%s: adId = %s", __PRETTY_FUNCTION__, adId.c_str());
    bridge_.call(kShowRewardedAd, adId);
}

void Self::onLoaded(const std::string& adId) {
    logger_.debug(__PRETTY_FUNCTION__);
    auto iter = rewardedAds_.find(adId);
    if (iter != rewardedAds_.cend()) {
        iter->second.weak->onLoaded();
    } else {
        // Mediation.
        assert(false);
    }
}

void Self::onFailedToLoad(const std::string& adId, const std::string& message) {
    logger_.debug(__PRETTY_FUNCTION__);
    auto iter = rewardedAds_.find(adId);
    if (iter != rewardedAds_.cend()) {
        iter->second.weak->onFailedToLoad(message);
    } else {
        // Mediation.
        assert(false);
    }
}

void Self::onFailedToShow(const std::string& adId, const std::string& message) {
    logger_.debug(__PRETTY_FUNCTION__);
    auto iter = rewardedAds_.find(adId);
    if (iter != rewardedAds_.cend()) {
        iter->second.weak->onFailedToShow(message);
    } else {
        // Mediation.
        assert(false);
    }
}

void Self::onClicked(const std::string& adId) {
    logger_.debug(__PRETTY_FUNCTION__);
    auto iter = rewardedAds_.find(adId);
    if (iter != rewardedAds_.cend()) {
        iter->second.weak->onClicked();
    } else {
        // Mediation.
        assert(false);
    }
}

void Self::onClosed(const std::string& adId, bool rewarded) {
    logger_.debug(__PRETTY_FUNCTION__);
    auto iter = rewardedAds_.find(adId);
    if (iter != rewardedAds_.cend()) {
        iter->second.weak->onClosed(rewarded);
    } else {
        // Mediation.
        assert(false);
        onMediationAdClosed(adId, rewarded);
    }
}

void Self::onMediationAdClosed(const std::string& adId, bool rewarded) {
    logger_.debug("%s: %s", __PRETTY_FUNCTION__,
                  core::toString(rewarded).c_str());
    if (rewardedAdDisplayer_->isProcessing()) {
        rewardedAdDisplayer_->resolve(rewarded ? IRewardedAdResult::Completed
                                               : IRewardedAdResult::Canceled);
        return;
    }
    assert(false);
}
} // namespace vungle
} // namespace ee
